### ISSUE TITLE
investigate using yamllint to avoid common yaml issues

### ISSUE THREAD

**Issue by linsun** [istio/istio#10956]
We can gate some of the simple yaml issues via yamllint:

https://pypi.org/project/yamllint/  

Example, Our team reported 2 simple issues on 1.0, which are fixed in 1.1 now:

https://github.com/istio/istio/issues/10954
https://github.com/istio/istio/issues/10955

**Comment by linsun** [comment]
Putting on hold, yamllint is written in GPL.

**Comment by linsun** [comment]
ok I ran through this tool locally on my machine with our istio install helm dir, there are lots of errors we need to fix.

@sdake do you see an issue with using a tool that is GPL for building/testing?  I don't... think it is okay as long as we don't use it at runtime.  cc @hklai @smawson

**Comment by linsun** [comment]
@irisdingbj do you have bandwidth to pick this up?  1) fix our current yaml to make yamllint happy 2) add this to our CI (circle/prow).

**Comment by sdake** [comment]
@linsun I am not an attorney, however, I have seen past large scale ASL2 projects that require permissive licenses permit the usage of a COPYLEFT license (such as GPL) for gating, but not for runtime.  Also depends on GPLv2 vs GPLv3.

I'd run it by the legal counsel - every company has a different interpretation of the GPL (and vastly different interpretations of GPLv3 specifically).

Cheers
-steve

**Comment by sdake** [comment]
fwiw, I am a big fan of linting 👍

**Comment by irisdingbj** [comment]
@linsun If we want to use yamllint we need decide which rules are used, Is the default relax one enough? Or do we want to any custom rules? Below is the default relax one:

rules:
  braces:
    level: warning
    max-spaces-inside: 1
  brackets:
    level: warning
    max-spaces-inside: 1
  colons:
    level: warning
  commas:
    level: warning
  comments: disable
  comments-indentation: disable
  document-start: disable
  empty-lines:
    level: warning
  hyphens:
    level: warning
  indentation:
    level: warning
    indent-sequences: consistent
  line-length:
    level: warning
    allow-non-breakable-inline-mappings: true
  truthy: disable

**Comment by linsun** [comment]
@irisdingbj default ok for me.

@kramvan1 any preference?

**Comment by irisdingbj** [comment]
Will go ahead with the default rule

**Comment by irisdingbj** [comment]
@linsun I have added yamlint into circi and make it part of the 'lint' target as below

```
  lint:
    <<: *defaults
    environment:
      KUBECONFIG: /go/src/istio.io/istio/.circleci/config
    resource_class: xlarge
    steps:
      - checkout
      - run: make sync
      - run: make lint
      - run: make yamllint
```
Also modified `.circleci/Dockerfile` to pull the `yamllint`
```
RUN  sudo apt-get -qqy install ruby ruby-dev rubygems build-essential autoconf libtool autotools-dev yamllint && \

```
But to make all this work, we need `docker push` a new circleci image into `istio/ci:go1.11-k8s1.10.4-helm2.7.2-minikube0.25`

Who should I contact to push the image up?

**Comment by howardjohn** [comment]
Helm has a `helm lint` specific to helm that may be useful as well.

**Comment by howardjohn** [comment]
@irisdingbj is this done?

**Comment by howardjohn** [comment]
Yamllint is in use now
